src/bigint-encoder.js

```diff
@@ -0,0 +1,97 @@
+/**
+ * Encode BigInt value from arbitrary provided arguments
+ *
+ * @param {Array<Number|BigInt|String>} parts - Slices to encode, where the
+ *    earlier elements represent higher bits in the final integer
+ * @param {64|128|256} size - Number of bits in the target integer type
+ * @param {Boolean} unsigned - Whether it's an unsigned integer
+ *
+ * @returns {BigInt}
+ */
+export function encodeBigIntFromBits(parts, size, unsigned) {
+  let result = 0n;
+  // check arguments length
+  if (parts.length && parts[0] instanceof Array) {
+    parts = parts[0];
+  }
+  const total = parts.length;
+  if (total === 1) {
+    try {
+      result = BigInt(parts[0]);
+      if (!unsigned) {
+        result = BigInt.asIntN(size, result);
+      }
+    } catch (e) {
+      throw new TypeError(`Invalid integer value: ${parts[0]}`)
+    }
+  } else {
+    const sliceSize = size / total;
+    if (sliceSize !== 32 && sliceSize !== 64 && sliceSize !== 128)
+      throw new TypeError('Invalid number of arguments')
+    // combine parts
+    for (let i = 0; i < total; i++) {
+      let part = BigInt.asUintN(sliceSize, BigInt(parts[i].valueOf()));
+      if (i > 0) { // shift if needed
+        part <<= BigInt(i * sliceSize);
+      }
+      result |= part;
+    }
+    if (!unsigned) { // clamp value to the requested size
+      result = BigInt.asIntN(size, result);
+    }
+  }
+  // check type
+  if (typeof result === 'bigint') {
+    // check boundaries
+    const [min, max] = calculateBigIntBoundaries(size, unsigned);
+    if (result >= min && result <= max)
+      return result;
+  }
+  // failed to encode
+  throw new TypeError(`Invalid ${formatIntName(size, unsigned)} value`);
+}
+
+/**
+ * @param {BigInt} value - Slices to encode
+ * @param {64|128|256} size - Number of bits in the source integer type
+ * @param {32|64|128} sliceSize - Number of parts
+ * @return {BigInt[]}
+ */
+export function sliceBigInt(value, size, sliceSize) {
+  if (typeof value !== 'bigint')
+    throw new TypeError('Invalid BigInt value');
+  const total = size / sliceSize;
+  if (total === 1)
+    return [value];
+  if (sliceSize < 32 || sliceSize > 128 || total !== 2 && total !== 4 && total !== 8)
+    throw new TypeError('Invalid slice size');
+  // prepare shift and mask
+  const shift = BigInt(sliceSize);
+  const mask = (1n << shift) - 1n;
+  // iterate shift and mask application
+  const result = new Array(total);
+  for (let i = 0; i < total; i++) {
+    if (i > 0) {
+      value >>= shift;
+    }
+    result[i] = BigInt.asIntN(sliceSize, value & mask); // clamp value
```

Comment:
Shouldn't this be `BigInt.asUintN` to get the "bytes"? Incidentally, I think `BigInt.asUintN` also masks based on the sliceSize, so you don't need both.

Reply:
You're right about the mask, but I'm not sure about the "signedness" here... @orbitlens?

Reply:
Don't remember why exactly I used `BigInt.asIntN`, but if I'm not mistaken, some tests were failing with `BigInt.asUintN` in this place. Probably because you can encode, say, `u64` from two `i32` parts, but other similar serialization libraries throw an error if you try to provide two `u32` parts to reconstruct `u64`. The convention is to use `i32`, not `u32` for constructing large integers. So it makes sense to always deconstruct an integer to signed smaller ints.

Reply:
fair enough, added a comment to that regard. 

we should _prooooobably_ modify `slice()` to use `asUintN` for `UnsignedHyper`, but given that's a new method anyway, it should be fine as-is too

Reply:
> we should prooooobably modify slice() to use asUintN for UnsignedHyper, but given that's a new method anyway, it should be fine as-is too

To my mind, the output of the `.slice()` method should be uniform for all integers regardless of the size and representation. I'd rather stick to outputting uniform _signed_ integers as a result of this method invocation, otherwise it may result in confusion. So using the `asIntN()` for all ints looks slightly more appealing to me.

Reply:
fine with me! you have a way better perspective given you are likely to be a user of these details so we can keep it as-is